services/src/test/java/io/grpc/services/HealthCheckingLoadBalancerFactoryTest.java

```diff
@@ -0,0 +1,1102 @@
+
+    // Concluded CONNECTING state
+    inOrder.verify(origLb).handleSubchannelState(
+        same(subchannel), eq(ConnectivityStateInfo.forNonError(CONNECTING)));
+
+    // Current health check RPC cancelled.
+    assertThat(serverCall.cancelled).isTrue();
+
+    // A second RPC is started immediately
+    assertThat(healthImpl.calls).hasSize(1);
+    serverCall = healthImpl.calls.poll();
+    // with the new service name
+    assertThat(serverCall.request).isEqualTo(makeRequest("FooService"));
+
+    verifyNoMoreInteractions(origLb);
+  }
+
+  @Test
+  public void serviceConfigChangesServiceNameWhenRetryPending() {
+    Attributes resolutionAttrs = attrsWithHealthCheckService("TeeService");
+    hcLbEventDelivery.handleResolvedAddressGroups(resolvedAddressList, resolutionAttrs);
+
+    verify(origLb).handleResolvedAddressGroups(same(resolvedAddressList), same(resolutionAttrs));
+    verifyNoMoreInteractions(origLb);
+
+    Subchannel subchannel = createSubchannel(0, Attributes.EMPTY);
+    assertThat(subchannel).isSameAs(subchannels[0]);
+    InOrder inOrder = inOrder(origLb);
+
+    hcLbEventDelivery.handleSubchannelState(subchannel, ConnectivityStateInfo.forNonError(READY));
+    inOrder.verify(origLb).handleSubchannelState(
+        same(subchannel), eq(ConnectivityStateInfo.forNonError(CONNECTING)));
+
+    HealthImpl healthImpl = healthImpls[0];
+    assertThat(healthImpl.calls).hasSize(1);
+    ServerSideCall serverCall = healthImpl.calls.poll();
+    assertThat(serverCall.cancelled).isFalse();
+    assertThat(serverCall.request).isEqualTo(makeRequest("TeeService"));
+
+    // Health check stream closed without responding.  Client in retry backoff.
+    assertThat(clock.getPendingTasks()).isEmpty();
+    serverCall.responseObserver.onCompleted();
+    assertThat(clock.getPendingTasks()).hasSize(1);
+    assertThat(healthImpl.calls).isEmpty();
+    inOrder.verify(origLb).handleSubchannelState(
+        same(subchannel),
+        unavailableStateWithMsg(
+            "Health-check stream was erroneously closed with " + Status.OK + " for 'TeeService'"));
+
+    // Service config returns with the same health check name.
+    hcLbEventDelivery.handleResolvedAddressGroups(resolvedAddressList, resolutionAttrs);
+    // It's delivered to origLb, but nothing else happens
+    inOrder.verify(origLb).handleResolvedAddressGroups(
+        same(resolvedAddressList), same(resolutionAttrs));
+    verifyNoMoreInteractions(origLb);
+    assertThat(clock.getPendingTasks()).hasSize(1);
+    assertThat(healthImpl.calls).isEmpty();
+
+    // Service config returns a different health check name.
+    resolutionAttrs = attrsWithHealthCheckService("FooService");
+    hcLbEventDelivery.handleResolvedAddressGroups(resolvedAddressList, resolutionAttrs);
+
+    // Concluded CONNECTING state
+    inOrder.verify(origLb).handleSubchannelState(
+        same(subchannel), eq(ConnectivityStateInfo.forNonError(CONNECTING)));
```

Comment:
Same here.

Reply:
Previous state is TRANSIENT_FAILURE for this case. I suppose you do want it to transition to CONNECTING.